File: src/skill_xray/capability.py
"""Manifest-scoped capability context; none of these states authorize behavior."""

from __future__ import annotations

import re
from copy import deepcopy
from dataclasses import dataclass, field

from .checks.code_lane import _governing_manifest, _manifest_index
from .checks.grants import (
    _EXECUTION_TOOLS,
    _NETWORK_TOOLS,
    declared_capabilities,
    denied_capabilities,
    effective_grants,
)

__all__ = ["CapabilityTriad", "build_triads"]

AXES = ("execution", "network")
_CLAIMS = {
    "execution": r"(?:runs?|executes?) (?:shell commands|commands|scripts)\b",
    "network": r"(?:uploads? (?:diagnostic logs|logs|data)|"
               r"sends? (?:diagnostic logs|logs|data) (?:over the network|via http)|"
               r"(?:access(?:es)?|uses?) the network|makes? (?:http|network) requests)\b",
}
_DENIALS = {
    "execution": r"(?:runs?|executes?) commands",
    "network": r"(?:access(?:es)?|uses?) the network",
}


def _unknown():
    return dict.fromkeys(AXES, "unknown")


@dataclass
class CapabilityTriad:
    manifest: str | None
    claimed: dict = field(default_factory=_unknown)
    declared: dict = field(default_factory=_unknown)
    observed: dict = field(default_factory=_unknown)
    evidence: list = field(default_factory=list)
    limitations: list = field(default_factory=list)


def _claims(manifest, triad):
    description = (manifest.frontmatter or {}).get("description")
    sources = []
    if isinstance(description, str):
        sources.append((description, manifest.frontmatter_key_lines.get("description")))
    if manifest.markdown is not None:
        lines = (manifest.text or "").splitlines()
        for start, end in manifest.markdown.prose_spans:
            block = lines[start - 1:min(end, len(lines))]
            if block and block[0].lower().startswith("this skill "):
                sources.append(("\n".join(block), start))
    states = {axis: set() for axis in AXES}
    for text, line in sources:
        evidence = []
        supported = True
        # A matching prefix cannot excuse an unsupported qualification elsewhere in the source.
        for statement in re.split(r"(?<=[.!])\s+", text.strip()):
            sentence = " ".join(statement.lower().split()).rstrip(".!")
            clauses = sentence.split(" and ")
            if len(statement) > 400 or (len(clauses) > 1 and re.search(
                    r"\b(?:does not|never)\b", sentence)):
                supported = False
                break
            for clause in clauses:
                clause = clause.removeprefix("this skill ")
                negative = re.match(r"^(?:does not|never)\s+", clause)
                action = clause[negative.end():] if negative else clause
                patterns = _DENIALS if negative else _CLAIMS
                matched = [axis for axis, pattern in patterns.items()
                           if re.fullmatch(pattern, action)]
                if not matched:
                    supported = False
                    break
                for axis in matched:
                    evidence.append({"path": manifest.rel, "line": line,
                                     "leg": "claimed", "capability": axis,
                                     "state": "denied" if negative else "present",
                                     "text": statement})
            if not supported:
                break
        if supported:
            for hit in evidence:
                states[hit["capability"]].add(hit["state"])
            triad.evidence.extend(evidence)
    for axis, values in states.items():
        if len(values) == 1:
            triad.claimed[axis] = next(iter(values))
        elif values:
            triad.limitations.append("conflicting-%s-claims" % axis)


def _declarations(manifest, triad):
    fm = manifest.frontmatter or {}
    fields = [key for key in ("allowed-tools", "disallowed-tools") if key in fm]
    grants = manifest.grants or ()
    if (any(code in {"frontmatter_parse_error", "grants_unparsed_shape"}
            for code, _ in manifest.diagnostics)
            or any(fm[key] is None or isinstance(fm[key], str) and not fm[key].strip()
                   or isinstance(fm[key], list) and any(
                       isinstance(item, str) and not item.strip() for item in fm[key])
                   for key in fields) or any(not grant.parsed for grant in grants)):
        triad.limitations.append("declaration-unparsed")
        return
    allowed = declared_capabilities(grants)
    # A denial of one URL or command does not deny the entire capability axis.
    denied = denied_capabilities(
        g for g in grants if (g.pattern or "").strip() in {"", "*", "**", ":*"})
    uncertain = set()
    for grant in effective_grants(grants):
        if grant.tool in _EXECUTION_TOOLS:
            # A command not recognized as network-capable is not proven network-free.
            uncertain.update(set(AXES) - declared_capabilities([grant]))
        elif grant.tool not in _NETWORK_TOOLS | {"Read", "Write", "Edit", "MultiEdit",
                                               "Glob", "Grep", "LS"}:
            uncertain.update(AXES)
    if uncertain:
        triad.limitations.append("declaration-capability-unknown")
    for axis in AXES:
        if axis in allowed:
            triad.declared[axis] = "present"
        elif axis not in uncertain and ("allowed-tools" in fm or axis in denied):
            triad.declared[axis] = "denied"
    for key in fields:
        triad.evidence.append({"path": manifest.rel, "leg": "declared", "field": key,
                               "line": manifest.frontmatter_key_lines.get(key)})


def build_triads(parsed, observations=(), coverage=()) -> dict[str, CapabilityTriad]:
    """Reuse governing manifests and validated engine observations; absence stays unknown.

    Claims recognize narrow English templates in descriptions and self-referential manifest
    prose. Other documents and arbitrary language remain unknown, not a positive disclosure.
    """
    manifests = _manifest_index(parsed)
    triads = {p.rel: CapabilityTriad(p.rel) for p in manifests.values()}
    observations = list(observations)
    coverage = list(coverage)
    for finding in coverage:
        if (finding.rule in {"preproc-inline-bang", "preproc-fenced-bang"}
                and finding.vector in {"SXV-001", "SXV-002"}
                and finding.evidence.get("command_sha256")):
            observations.append({"path": finding.path, "line": finding.line,
                                 "column": finding.column, "capability": "execution",
                                 "state": "present", "analyzer": "preproc-ir",
                                 "rule": finding.rule, "vector": finding.vector,
                                 "command_sha256": finding.evidence["command_sha256"]})
    for observation in observations:
        path, axis = observation.get("path", ""), observation.get("capability")
        manifest = _governing_manifest(manifests, path)
        key = manifest.rel if manifest else ""
        triad = triads.setdefault(key, CapabilityTriad(key or None))
        triad.evidence.append(deepcopy(observation))
        if axis in AXES and observation.get("state") == "present":
            triad.observed[axis] = "present"
        else:
            triad.limitations.append(observation.get("reason", "observation-unvalidated"))
    for manifest in manifests.values():
        triad = triads[manifest.rel]
        _claims(manifest, triad)
        _declarations(manifest, triad)
        if manifest.text is None:
            triad.limitations.append("manifest-unread")
    for finding in coverage:
        if finding.vector:
            continue
        manifest = _governing_manifest(manifests, finding.path) if finding.path else None
        affected = [triads[manifest.rel]] if manifest else list(triads.values())
        for triad in affected:
            triad.limitations.append(finding.rule)
    for triad in triads.values():
        triad.limitations = sorted(set(triad.limitations))
    return dict(sorted(triads.items()))

File: internal/capability/capability.go
// Package capability builds the manifest-scoped capability context (Python capability.py): for
// every governing SKILL.md, what its prose claims, what its tool grants declare and what the
// engines observed in the code it governs. None of these states authorise behaviour.
package capability

import (
	"regexp"
	"slices"
	"strings"
	"unicode/utf8"

	"github.com/traceforce/skill-xray/internal/checks"
	"github.com/traceforce/skill-xray/internal/findings"
	"github.com/traceforce/skill-xray/internal/grants"
	"github.com/traceforce/skill-xray/internal/parse"
	"github.com/traceforce/skill-xray/internal/pytext"
)

var axes = []string{"execution", "network"}

// Triad is CapabilityTriad; each axis map holds "unknown", "present" or "denied".
type Triad struct {
	Manifest    *string           `json:"manifest"`
	Claimed     map[string]string `json:"claimed"`
	Declared    map[string]string `json:"declared"`
	Observed    map[string]string `json:"observed"`
	Evidence    []map[string]any  `json:"evidence"`
	Limitations []string          `json:"limitations"`
}

func newTriad(manifest string) *Triad {
	t := &Triad{Claimed: map[string]string{}, Declared: map[string]string{}, Observed: map[string]string{},
		Evidence: []map[string]any{}, Limitations: []string{}}
	for _, axis := range axes {
		t.Claimed[axis], t.Declared[axis], t.Observed[axis] = "unknown", "unknown", "unknown"
	}
	if manifest != "" {
		t.Manifest = &manifest
	}
	return t
}

var (
	// _CLAIMS and _DENIALS, matched whole; Python's trailing \b is implied by $ after a letter.
	claims = map[string]*regexp.Regexp{
		"execution": regexp.MustCompile(`^(?:runs?|executes?) (?:shell commands|commands|scripts)$`),
		"network": regexp.MustCompile(`^(?:uploads? (?:diagnostic logs|logs|data)|` +
			`sends? (?:diagnostic logs|logs|data) (?:over the network|via http)|` +
			`(?:access(?:es)?|uses?) the network|makes? (?:http|network) requests)$`),
	}
	denials = map[string]*regexp.Regexp{
		"execution": regexp.MustCompile(`^(?:runs?|executes?) commands$`),
		"network":   regexp.MustCompile(`^(?:access(?:es)?|uses?) the network$`),
	}
	// Python's \b is Unicode-aware; an explicit non-word rune or the text edge replaces it.
	qualifierRE = regexp.MustCompile(`(?i)^(?:this (?:is|was)|that|these|those|but|however|unless|except|only|not|never|does not)(?:[^\p{L}\p{N}_]|$)`)
	exampleRE   = regexp.MustCompile(`(?i)^(?:(?:for )?example|the following example)(?:[^\p{L}\p{N}_]|$)`)
	negationRE  = regexp.MustCompile(`(?:^|[^\p{L}\p{N}_])(?:does not|never)(?:[^\p{L}\p{N}_]|$)`)
	negativeRE  = regexp.MustCompile(`^(?:does not|never)\s+`)
	// re.split(r"(?<=[.!])\s+") without the lookbehind: a statement ends after the punctuation
	// and the next starts after the whitespace run.
	statementEndRE = regexp.MustCompile(`[.!]` + pytext.Space + `+`)
	// Tools whose grants leave both axes certain (the read/write set of _declarations).
	fileTools = map[string]bool{"Read": true, "Write": true, "Edit": true, "MultiEdit": true, "Glob": true, "Grep": true, "LS": true}
)

func statements(text string) []string {
	out, pos := []string{}, 0
	for _, m := range statementEndRE.FindAllStringIndex(text, -1) {
		out = append(out, text[pos:m[0]+1])
		pos = m[1]
	}
	return append(out, text[pos:])
}

// sub is Python's lines[a:b] for 0 <= a: both bounds clamp to the slice.
func sub(lines []string, a, b int) []string {
	a = min(a, len(lines))
	return lines[a:max(min(b, len(lines)), a)]
}

// claims is _claims: the description and every self-referential ("This skill ...") top-level
// paragraph are read as complete statements; one unrecognised statement voids its whole source.
func (t *Triad) claims(m *parse.Artifact) {
	type source struct {
		text  string
		line  int
		prose bool
	}
	var sources []source
	if d, ok := m.Frontmatter["description"].(string); ok {
		sources = append(sources, source{d, m.FrontmatterKeyLines["description"], false})
	}
	if m.Markdown != nil {
		var text string
		if m.Text != nil {
			text = *m.Text
		}
		lines := pytext.SplitLines(text)
		var spans []parse.Span
		for _, s := range m.Markdown.ParagraphSpans {
			qualifier := qualifierRE.MatchString(pytext.Strip(lines[s.Start-1]))
			if n := len(spans); n > 0 && (qualifier || exampleRE.MatchString(pytext.Strip(lines[spans[n-1].Start-1]))) &&
				!slices.ContainsFunc(sub(lines, spans[n-1].End, s.Start-1), func(l string) bool { return pytext.Strip(l) != "" }) {
				spans[n-1].End = s.End
			} else {
				spans = append(spans, s)
			}
		}
		for _, s := range spans {
			if block := sub(lines, s.Start-1, s.End); len(block) > 0 && strings.HasPrefix(pytext.Lower(block[0]), "this skill ") {
				sources = append(sources, source{strings.Join(block, "\n"), s.Start, true})
			}
		}
	}
	states := map[string]string{}
	for _, src := range sources {
		var evidence []map[string]any
		supported, position, line := true, 0, src.line
		for _, statement := range statements(pytext.Strip(src.text)) {
			start := position + strings.Index(src.text[position:], statement)
			statementLine := line
			if src.prose {
				statementLine = line + strings.Count(src.text[position:start], "\n")
				line = statementLine + strings.Count(statement, "\n")
			}
			position = start + len(statement)
			sentence := strings.TrimRight(strings.Join(pytext.Fields(pytext.Lower(statement)), " "), ".!")
			clauses := strings.Split(sentence, " and ")
			if utf8.RuneCountInString(statement) > 400 || len(clauses) > 1 && negationRE.MatchString(sentence) {
				supported = false
				break
			}
			for _, clause := range clauses {
				clause = strings.TrimPrefix(clause, "this skill ")
				action, state, patterns := clause, "present", claims
				if neg := negativeRE.FindStringIndex(clause); neg != nil {
					action, state, patterns = clause[neg[1]:], "denied", denials
				}
				matched := false
				for _, axis := range axes {
					if patterns[axis].MatchString(action) {
						matched = true
						evidence = append(evidence, map[string]any{"path": m.Rel, "line": statementLine,
							"leg": "claimed", "capability": axis, "state": state, "text": statement})
					}
				}
				if !matched {
					supported = false
					break
				}
			}
			if !supported {
				break
			}
		}
		if supported {
			for _, hit := range evidence {
				axis, state := hit["capability"].(string), hit["state"].(string)
				if prev, seen := states[axis]; seen && prev != state {
					state = "conflict"
				}
				states[axis] = state
			}
			t.Evidence = append(t.Evidence, evidence...)
		}
	}
	for _, axis := range axes {
		switch s := states[axis]; s {
		case "":
		case "conflict":
			t.Limitations = append(t.Limitations, "conflicting-"+axis+"-claims")
		default:
			t.Claimed[axis] = s
		}
	}
}

// blankValue is the unparsed-declaration test of _declarations: a null, a blank string or a
// list holding a blank string.
func blankValue(v any) bool {
	switch v := v.(type) {
	case nil:
		return true
	case string:
		return pytext.Strip(v) == ""
	case []any:
		return slices.ContainsFunc(v, func(item any) bool { s, ok := item.(string); return ok && pytext.Strip(s) == "" })
	}
	return false
}

// declarations is _declarations: explicit states only from fully parsed grants; anything
// unparsed or unrecognised is a limitation, never a denial.
func (t *Triad) declarations(m *parse.Artifact) {
	fields := slices.DeleteFunc([]string{"allowed-tools", "disallowed-tools"}, func(k string) bool { _, ok := m.Frontmatter[k]; return !ok })
	if slices.ContainsFunc(m.Diagnostics, func(d parse.Diagnostic) bool {
		return d.Code == "frontmatter_parse_error" || d.Code == "grants_unparsed_shape"
	}) || slices.ContainsFunc(fields, func(k string) bool { return blankValue(m.Frontmatter[k]) }) ||
		slices.ContainsFunc(m.Grants, func(g parse.Grant) bool { return !g.Parsed }) {
		t.Limitations = append(t.Limitations, "declaration-unparsed")
		return
	}
	allowed, denied := grants.Declared(m.Grants), grants.Denied(m.Grants)
	uncertain := map[string]bool{}
	for _, g := range grants.Effective(m.Grants) {
		switch {
		case grants.ExecutionTools[g.Tool]: // a command not proven network-capable is not proven network-free
			one := grants.Declared([]parse.Grant{g})
			for _, axis := range axes {
				if !one[axis] {
					uncertain[axis] = true
				}
			}
		case !grants.NetworkTools[g.Tool] && !fileTools[g.Tool]:
			for _, axis := range axes {
				uncertain[axis] = true
			}
		}
	}
	if len(uncertain) > 0 {
		t.Limitations = append(t.Limitations, "declaration-capability-unknown")
	}
	_, hasAllowed := m.Frontmatter["allowed-tools"]
	for _, axis := range axes {
		if allowed[axis] {
			t.Declared[axis] = "present"
		} else if !uncertain[axis] && (hasAllowed || denied[axis]) {
			t.Declared[axis] = "denied"
		}
	}
	for _, k := range fields {
		t.Evidence = append(t.Evidence, map[string]any{"path": m.Rel, "leg": "declared", "field": k, "line": m.FrontmatterKeyLines[k]})
	}
}

// Build is build_triads: one triad per governing manifest, plus "" for observations no
// manifest governs, keyed by manifest rel. Observations are the engines' records;
// coverage is the raw finding list, whose preprocessing hits become execution observations and
// whose vector-less gaps become limitations of the manifests they touch.
func Build(p *parse.Package, observations []map[string]any, coverage []findings.Finding) map[string]*Triad {
	manifests := parse.ManifestIndex(p)
	triads := map[string]*Triad{}
	for _, m := range manifests {
		triads[m.Rel] = newTriad(m.Rel)
	}
	observations = slices.Clone(observations)
	for _, f := range coverage {
		sha, _ := f.Evidence["command_sha256"].(string)
		if (f.Rule == "preproc-inline-bang" || f.Rule == "preproc-fenced-bang") &&
			(f.Vector == "SXV-001" || f.Vector == "SXV-002") && sha != "" {
			observations = append(observations, map[string]any{"path": f.Path, "line": f.Line,
				"column": f.Column, "capability": "execution", "state": "present",
				"analyzer": "preproc-ir", "rule": f.Rule, "vector": f.Vector, "command_sha256": sha})
		}
	}
	for _, obs := range observations {
		path, _ := obs["path"].(string)
		key := ""
		if m := parse.GoverningManifest(manifests, path); m != nil {
			key = m.Rel
		}
		t := triads[key]
		if t == nil {
			t = newTriad(key)
			triads[key] = t
		}
		t.Evidence = append(t.Evidence, pytext.DeepCopy(obs).(map[string]any))
		if axis, _ := obs["capability"].(string); slices.Contains(axes, axis) && obs["state"] == "present" {
			t.Observed[axis] = "present"
		} else if reason, ok := obs["reason"].(string); ok {
			t.Limitations = append(t.Limitations, reason)
		} else {
			t.Limitations = append(t.Limitations, "observation-unvalidated")
		}
	}
	for _, m := range manifests {
		t := triads[m.Rel]
		t.claims(m)
		t.declarations(m)
		if m.Text == nil {
			t.Limitations = append(t.Limitations, "manifest-unread")
		}
	}
	for _, f := range coverage {
		if f.Vector != "" || checks.IsInventoryNote(f.Vector, f.Rule, f.Severity, f.Evidence) {
			continue
		}
		var m *parse.Artifact
		if f.Path != "" {
			m = parse.GoverningManifest(manifests, f.Path)
		}
		for _, t := range triads {
			if m == nil || t == triads[m.Rel] {
				t.Limitations = append(t.Limitations, f.Rule)
			}
		}
	}
	for _, t := range triads {
		slices.Sort(t.Limitations)
		t.Limitations = slices.Compact(t.Limitations)
	}
	return triads
}
